### ISSUE TITLE
git sync: new strategy "ff-only"

### ISSUE THREAD

**Issue by kevgo** [git-town/git-town#4104]
When this sync strategy is selected and

- a branch has no local commits: sync the branch via `git merge --ff-only`
- a branch has an unpushed local commit: don't sync the branch, print a notification to the user

replaces #3914
 
alternative name: `ff-only`

_Originally posted by @sergej-koscejev in https://github.com/git-town/git-town/issues/3914#issuecomment-2400404037_

> Can there be a setting like fast-forward-only? My perennial branches are protected on GitHub and I shouldn’t have any commits on the perennial branches directly. I also don’t want git town to push them but that might be another discussion.

**Comment by kevgo** [comment]
Open question: which alternative implementation of this idea is better?

#### Alternative 1: `no-local-commits` sync-strategy

- if the local branch is in sync or behind its tracking branch: runs `git merge origin/branch-name --ff-only`
- if not: does not sync the branch and queues a message to the user to be printed when the sync command is done

pro:
- directly expresses the intent ("I don't want Git Town to push any local commits on this branch") and what Git Town does (it won't sync local commits on this branch)
- helpful error message at the end of the sync, e.g. "branch XYZ should not contain local commits"

con:
- name feels a bit out of place compared to the names of the other sync strategies (`merge`, `rebase`, `compress`) and not idiomatic for a Git workflow tool

#### Alternative 2: `ff-only` sync-strategy

- always runs `git merge origin/branch-name --ff-only`
- if the local branch is in sync or behind its tracking branch: does the right thing
- if not: exits with a Git error message like `fatal: Not possible to fast-forward, aborting.`

pro:
- name is closer to the underlying Git domain model
- this might be useful for other scenarios

con:
- name of this sync strategy doesn't communicate that Git Town won't push local commits
- when there is a local commit, the error message given to the user is not helpful
- This encourages/forces the user to move the commit to another branch while in the middle of an active sync. They could run `git town skip` to skip syncing that branch and move the commit when the sync is done. But then they need to remember that the `skip` command exists, and they also need to remember to move the commit when the sync is done. Overall this is a doable user experience, but it isn't the best user experience for such a common case.

**Comment by sergej-koscejev** [comment]
You could also call it ff-only but do what is described by no-local-commits, i.e. no need to run git merge --ff-only just to have it fail with an unhelpful error message if you can detect the problem in advance and fail with a helpful error message (which could suggest `git town skip` to the user).

**Comment by kevgo** [comment]
An emerging pattern is:

- sync the tracking branch with `git merge origin/branch --ff-only`
- no syncing with the parent branch --> this sync strategy is only allowed for main, perennial, and observed branches

Given this, the name `ff-only` seems doable and more appropriate.

**Comment by kevgo** [comment]
@sergej-koscejev I agree, came to the same conclusion 👍 Git Town knows the sync status of branches, i.e. whether it is ahead or behind its tracking branch.

**Comment by kevgo** [comment]
This is now available as part of [Git Town 17.3](https://github.com/git-town/git-town/releases/tag/v17.3.0). Enjoy!
